### ISSUE TITLE
new lint: using `let mut a = a` at start of function

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#5102]
I inherited a codebase from someone new to Rust and there were lots of function definitions that looked like this:

```rust
fn f(a: Vec<u8>) {
    let mut a = a;
    // rest of code
}
```

It would be nice if there were a lint saying this would be better as `fn f(mut a: Vec<u8>)` instead.

**Comment by sonng** [comment]
Looks like a good first issue. Could this be assigned to me please?

Thanks

**Comment by sonng** [comment]
Hey guys,

I'm on the last leg of the work (`cargo test`) and making sure everything is fine. I noticed that while running that, our new lint gets triggered and causes an error, causing the whole thing to fail.

My question is, how do I turn this from an error to a warning?
or alternatively, I could update the problematic code and make them fn parameters mutable.

Cheers

**Comment by flip1995** [comment]
Can you open a WIP PR, so I can have a look at the code and the dogfood error?

**Comment by sonng** [comment]
I'm happy for any feedback on the overall change too.

**Comment by brightly-salty** [comment]
I'm willing to attempt to implement this and submit a PR if it would be accepted if @sonng isn't working on it. I expect to restart (to avoid merge conflicts) using their PR as a guide/inspiration?

**Comment by flip1995** [comment]
@brightly-salty Thanks! Yes you can use they're PR as inspiration. I think my comments on the PR should also help you with dos and don'ts.

**Comment by max-niederman** [comment]
I think this should probably be generalized to redefined locals as well.

**Comment by maxcnunes** [comment]
Is this issue still relevant? It seems https://github.com/rust-lang/rust-clippy/pull/10885 covered what have been requested here and we could close it.
